components/sensor/ultrasonic/ultrasonic.go

```diff
@@ -134,96 +136,53 @@ func (s *Sensor) namedError(err error) error {
 	)
 }
 
-func (s *Sensor) startUpdateLoop(ctx context.Context) {
-	s.activeBackgroundWorkers.Add(1)
-	rdkutils.ManagedGo(
-		func() {
-			s.echoInterrupt.AddCallback(s.intChan)
-			defer s.echoInterrupt.RemoveCallback(s.intChan)
-			for {
-				select {
-				case <-s.cancelCtx.Done():
-					return
-				case <-s.readingChan:
-					// a call to Readings has occurred so we request a distance
-					// reading from the sensor and send it to the distance channel
-					if err := s.measureDistance(ctx); err != nil {
-						s.errChan <- err
-					}
-				// we must consume any signals from the interrupt that occur
-				// outside of a call to readings, otherwise we will potentially
-				// block callback code for *all* interrupts (see the implementation
-				// of pigpioInterruptCallback in components/board/pi/impl/board.go)
-				case <-s.intChan:
-				}
-			}
-		},
-		s.activeBackgroundWorkers.Done,
-	)
-}
-
-func (s *Sensor) measureDistance(ctx context.Context) error {
+// Readings returns the calculated distance.
+func (s *Sensor) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
 	// we send a high and a low to the trigger pin 10 microseconds
 	// apart to signal the sensor to begin sending the sonic pulse
 	if err := s.triggerPin.Set(ctx, true, nil); err != nil {
-		return s.namedError(errors.Wrap(err, "ultrasonic cannot set trigger pin to high"))
+		return nil, s.namedError(errors.Wrap(err, "ultrasonic cannot set trigger pin to high"))
 	}
 	rdkutils.SelectContextOrWait(ctx, time.Microsecond*10)
```

Comment:
My work with the oscilloscope and software PWMs yesterday (https://github.com/viamrobotics/rdk/pull/1881) suggests that this is likely to wait at least 1 millisecond before it wakes up again (at least, on a Jetson Orin, which I think of as a high-end board). How confident are you that you really wait for only 10 microseconds? How much of a bummer would it be to wait for much longer than that?

Reply:
I think 10 microseconds is just a minimum amount of wait time given that readings still come through perfectly fine

Reply:
@penguinland  i believe the sensor spec calls for a 10us pulse on the trigger pin. i don't know how much tolerance it has in either direction, but "waiting longer" may not be better here.